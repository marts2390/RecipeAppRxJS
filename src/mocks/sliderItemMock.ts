const sliderItemMock = [
  {
    idCategory: '1',
    strCategory: 'Beef',
    strCategoryDescription: 'Beef is the culinary name for meat from cattle, particularly skeletal muscle. Humans have been eating beef since prehistoric times.[1] Beef is a source of high-quality protein and essential nutrients.',
    strCategoryThumb: 'https://www.themealdb.com/images/category/beef.png',
  },
  {
    idCategory: '2',
    strCategory: 'Chicken',
    strCategoryDescription: 'Chicken is a type of domesticated fowl, a subspecies of the red junglefowl. It is one of the most common and widespread domestic animals, with a total population of more than 19 billion as of 2011.[1] Humans commonly keep chickens as a source of food (consuming both their meat and eggs) and, more rarely, as pets.',
    strCategoryThumb: 'https://www.themealdb.com/images/category/chicken.png',
  },
  {
    idCategory: '3',
    strCategory: 'Dessert',
    strCategoryDescription: 'Dessert is a course that concludes a meal. The course usually consists of sweet foods, such as confections dishes or fruit, and possibly a beverage such as dessert wine or liqueur, however in the United States it may include coffee, cheeses, nuts, or other savory items regarded as a separate course elsewhere. In some parts of the world, such as much of central and western Africa, and most parts of China, there is no tradition of a dessert course to conclude a meal.\r\n\r\nThe term dessert can apply to many confections, such as biscuits, cakes, cookies, custards, gelatins, ice creams, pastries, pies, puddings, and sweet soups, and tarts. Fruit is also commonly found in dessert courses because of its naturally occurring sweetness. Some cultures sweeten foods that are more commonly savory to create desserts.',
    strCategoryThumb: 'https://www.themealdb.com/images/category/dessert.png',
  }
]

export default sliderItemMock
