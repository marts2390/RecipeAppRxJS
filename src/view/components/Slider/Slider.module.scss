@import 'src/view/styles/variables.scss';

.categoryCard {
  position: relative;
  margin: 0 1em;
  background-position: center;
  background-size: cover;
  border-radius: 5px;
  height: 7em;
  overflow: hidden;

  .cardOverlay {
    transition: background-color 0.3s ease;
    position: absolute;
    top: 0;
    left: 0;
    height: 100%;
    width: 100%;
    background-color: rgba(black, 0.5);
    display: flex;
    align-items: center;
    justify-content: center;

    &:hover {
      background-color: rgba(black, 0.35);
      cursor: pointer;

      .title {
        font-size: $font-md;
        letter-spacing: 2px;
        font-weight: 300;
      }
    }
  }

  .title {
    color: white;
    transition: 0.3s ease;
    font-size: $font-sm;
    font-weight: 300;
  }
}

.categorySlider {
  display: flex;
}

.controlButton {
  background: none;
  border: none;
  position: absolute;
  top: 1.5em;
  font-size: 2em;

  &:focus {
    outline: none;
  }
}

.slider {
  position: relative;
}

.nextButton {
  right: -20px;
}

.backButton {
  left: -20px;
}
