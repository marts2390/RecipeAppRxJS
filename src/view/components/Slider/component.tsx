import React, { FC, useState, useEffect } from 'react'
import {
  CarouselProvider, Slider, Slide, ButtonBack, ButtonNext
} from 'pure-react-carousel'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faChevronLeft, faChevronRight } from '@fortawesome/free-solid-svg-icons'
import cn from 'classnames'

import { ICategoriesDataModel } from 'models/categoriesDataModel'

import 'pure-react-carousel/dist/react-carousel.es.css'
import styles from './Slider.module.scss'

interface ISliderComponentProps {
  sliderItems:ICategoriesDataModel[]
  sliderAction: (category: string) => void;
}

const SliderComponent: FC<ISliderComponentProps> = ({ sliderItems, sliderAction }) => {
  const backButtonClasses = cn(styles.backButton, styles.controlButton)
  const nextButtonClasses = cn(styles.nextButton, styles.controlButton)

  const [windowWidth, setWindowWidth] = useState(window.innerWidth)

  useEffect(() => {
    const handleWindowResize = () => {
      setWindowWidth(window.innerWidth)
    }

    window.addEventListener('resize', handleWindowResize)

    return () => {
      window.removeEventListener('resize', handleWindowResize)
    }
  }, [])

  let slidesToShow = 1
  if (windowWidth > 400 && windowWidth < 767) {
    slidesToShow = 2
  } else if (windowWidth > 768 && windowWidth < 999) {
    slidesToShow = 4
  } else if (windowWidth > 1000 && windowWidth < 1599) {
    slidesToShow = 6
  } else if (windowWidth > 1600) {
    slidesToShow = 10
  }

  return (
    <CarouselProvider
      naturalSlideWidth={ 50 }
      naturalSlideHeight={ 40 }
      totalSlides={ sliderItems?.length }
      visibleSlides={ slidesToShow }
      step={ 1 }
      className={ styles.slider }
      isIntrinsicHeight
    >
      <Slider>
        {sliderItems?.map((item, index) => (
          <Slide
            index={ index }
            key={ item.idCategory }
          >
            <div
              className={ styles.categoryCard }
              style={ { backgroundImage: `url(${item.strCategoryThumb})` } }
            >
              <div
                role="button"
                onKeyDown={ (e) => e.key === 'Enter' && sliderAction(item.strCategory) }
                className={ styles.cardOverlay }
                onClick={ () => sliderAction(item.strCategory) }
              >
                <h3 className={ styles.title }>{item.strCategory}</h3>
              </div>
            </div>
          </Slide>
        ))}
      </Slider>
      <ButtonBack className={ backButtonClasses }>
        <FontAwesomeIcon icon={ faChevronLeft } />
      </ButtonBack>
      <ButtonNext className={ nextButtonClasses }>
        <FontAwesomeIcon icon={ faChevronRight } />
      </ButtonNext>
    </CarouselProvider>
  )
}

export default SliderComponent
