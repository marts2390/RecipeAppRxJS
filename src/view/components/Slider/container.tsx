import React, { FC } from 'react'
import UC from 'actions'

import { ICategoriesDataModel } from 'models/categoriesDataModel'

import SliderComponent from './component'

interface SliderContainerProps {
  sliderItems:ICategoriesDataModel[]
  sliderAction: (category: string) => void;
}

const SliderContainer:FC<SliderContainerProps> = ({ sliderItems, sliderAction }) => (
  <SliderComponent
    sliderItems={ sliderItems }
    sliderAction={ sliderAction }
  />
)

export default SliderContainer
