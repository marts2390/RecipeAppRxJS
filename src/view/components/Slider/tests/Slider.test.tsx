import React from 'react'
import { shallow } from 'enzyme'
import sliderItemMock from '../../../../mocks/sliderItemMock'
import Slider from '../component'

let wrapper
let props

beforeEach(() => {
  props = {
    sliderItems: sliderItemMock,
    sliderAction: jest.fn()
  }

  wrapper = shallow(<Slider { ...props } />)
})

test('Slider is rendered', () => {
  expect(wrapper).toMatchSnapshot()
})

test('Slider is rendered with no items', () => {
  wrapper.setProps({ sliderItems: [] })
  expect(wrapper).toMatchSnapshot()
})

test('should call sliderAction on click', () => {
  const sliderItem = wrapper.find({ 'data-testid': 'recipe-card-1' })
  wrapper.find({ 'data-testid': 'recipe-card-1' }).simulate('click')
  sliderItem.prop('onClick')
  expect(props.sliderAction).toHaveBeenCalledWith(props.sliderItems[1].strCategory)
})

test('should call sliderAction on keyDown', () => {
  const sliderItem = wrapper.find({ 'data-testid': 'recipe-card-1' })
  sliderItem.prop('onKeyDown')({ key: 'Enter' })
  expect(props.sliderAction).toHaveBeenCalledWith(props.sliderItems[1].strCategory)
})
