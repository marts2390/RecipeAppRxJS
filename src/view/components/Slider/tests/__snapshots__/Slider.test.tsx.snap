// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider is rendered 1`] = `
<i
  className="slider"
  currentSlide={0}
  disableAnimation={false}
  disableKeyboard={false}
  dragEnabled={true}
  dragStep={1}
  hasMasterSpinner={false}
  infinite={false}
  interval={5000}
  isIntrinsicHeight={true}
  isPageScrollLocked={false}
  isPlaying={false}
  lockOnWindowScroll={false}
  naturalSlideHeight={40}
  naturalSlideWidth={50}
  orientation="horizontal"
  playDirection="forward"
  step={1}
  tag="div"
  totalSlides={3}
  touchEnabled={true}
  visibleSlides={6}
>
  <i>
    <i
      index={0}
      key="1"
    >
      <div
        className="categoryCard"
        style={
          Object {
            "backgroundImage": "url(https://www.themealdb.com/images/category/beef.png)",
          }
        }
      >
        <div
          className="cardOverlay"
          data-testid="recipe-card-0"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
        >
          <h3
            className="title"
          >
            Beef
          </h3>
        </div>
      </div>
    </i>
    <i
      index={1}
      key="2"
    >
      <div
        className="categoryCard"
        style={
          Object {
            "backgroundImage": "url(https://www.themealdb.com/images/category/chicken.png)",
          }
        }
      >
        <div
          className="cardOverlay"
          data-testid="recipe-card-1"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
        >
          <h3
            className="title"
          >
            Chicken
          </h3>
        </div>
      </div>
    </i>
    <i
      index={2}
      key="3"
    >
      <div
        className="categoryCard"
        style={
          Object {
            "backgroundImage": "url(https://www.themealdb.com/images/category/dessert.png)",
          }
        }
      >
        <div
          className="cardOverlay"
          data-testid="recipe-card-2"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
        >
          <h3
            className="title"
          >
            Dessert
          </h3>
        </div>
      </div>
    </i>
  </i>
  <i
    className="backButton controlButton"
  >
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            320,
            512,
            Array [],
            "f053",
            "M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z",
          ],
          "iconName": "chevron-left",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      swapOpacity={false}
      symbol={false}
      title=""
      transform={null}
    />
  </i>
  <i
    className="nextButton controlButton"
  >
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            320,
            512,
            Array [],
            "f054",
            "M285.476 272.971L91.132 467.314c-9.373 9.373-24.569 9.373-33.941 0l-22.667-22.667c-9.357-9.357-9.375-24.522-.04-33.901L188.505 256 34.484 101.255c-9.335-9.379-9.317-24.544.04-33.901l22.667-22.667c9.373-9.373 24.569-9.373 33.941 0L285.475 239.03c9.373 9.372 9.373 24.568.001 33.941z",
          ],
          "iconName": "chevron-right",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      swapOpacity={false}
      symbol={false}
      title=""
      transform={null}
    />
  </i>
</i>
`;

exports[`Slider is rendered with no items 1`] = `
<i
  className="slider"
  currentSlide={0}
  disableAnimation={false}
  disableKeyboard={false}
  dragEnabled={true}
  dragStep={1}
  hasMasterSpinner={false}
  infinite={false}
  interval={5000}
  isIntrinsicHeight={true}
  isPageScrollLocked={false}
  isPlaying={false}
  lockOnWindowScroll={false}
  naturalSlideHeight={40}
  naturalSlideWidth={50}
  orientation="horizontal"
  playDirection="forward"
  step={1}
  tag="div"
  totalSlides={0}
  touchEnabled={true}
  visibleSlides={6}
>
  <i />
  <i
    className="backButton controlButton"
  >
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            320,
            512,
            Array [],
            "f053",
            "M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z",
          ],
          "iconName": "chevron-left",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      swapOpacity={false}
      symbol={false}
      title=""
      transform={null}
    />
  </i>
  <i
    className="nextButton controlButton"
  >
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            320,
            512,
            Array [],
            "f054",
            "M285.476 272.971L91.132 467.314c-9.373 9.373-24.569 9.373-33.941 0l-22.667-22.667c-9.357-9.357-9.375-24.522-.04-33.901L188.505 256 34.484 101.255c-9.335-9.379-9.317-24.544.04-33.901l22.667-22.667c9.373-9.373 24.569-9.373 33.941 0L285.475 239.03c9.373 9.372 9.373 24.568.001 33.941z",
          ],
          "iconName": "chevron-right",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      swapOpacity={false}
      symbol={false}
      title=""
      transform={null}
    />
  </i>
</i>
`;
